design/async-commit/spec.md

```diff
@@ -68,15 +73,21 @@ message LockInfo {
+
+The client must set `use_async_commit` and/or `try_one_pc` in the `PrewriteRequest`. If using async commit, then the `secondaries` field should contain the keys of all keys locked in the transaction except the primary key. If the user wants external consistency, a new timestamp is fetched from PD and used as the `min_commit_ts`. Otherwise, it is set to the maximum of the transaction's start and 'for update' timestamps (if we get a timestamp from PD it should be greater than both of those).
 
-TODO prewrite and commit
+When the client receives the `PrewriteResponse`, it should check the `one_pc_commit_ts` to see if the server used 1pc: if it is non-zero, then 1pc was used and that timestamp should be returned to the user with a success message. If it is non-zero, then 1pc failed and the client should continue with the 2pc protocol.
 
-TODO resolve
+The client calculates the overall minimum commit timestamp from the maximum of the `min_commit_ts` of each response. If any `min_commit_ts` is zero, then async commit failed and the client should fallback to 2pc (see below). That overall minimum commit timestamp is returned to the user as the commit timestamp of the transaction with a success message.
 
-TODO fallback
+With async 2pc, the primary key is committed asynchronously along with the secondary keys. This is done by sending the overall minimum commit timestamp as the commit timestamp in a commit message to each store.
 
-TODO schema check
+If a transaction (async or sync) encounters a lock belonging to an async transaction, there is a new way to resolve the lock. The client first sends a `CheckTransactionStatusRequest` to the primary lock (this request should be retried since it is possible that a secondary lock exists before its primary lock does). Once the lock's TTL has expired, the response will indicate if the lock has been committed, rolled back, or is still locked. If it is still locked, then the client will use the list of secondary locks in the response. The client will send a `CheckSecondaryLocksRequest` to each lock, the server will return the state of each lock and (if appropriate) a commit timestamp. The client can then resolve locks; if any lock was rolled back or committed, then all locks should be rolled back or committed. If all keys are locked (or have never been locked) then the client can commit the transaction using the maximum of all lock's `min_commit_ts`.
+
+If using async commit or 1pc, the server may not be able to use these techniques to complete the transaction. In this case, the client falls back to using plain 2pc. Then the client will set `force_sync_commit` to true when requesting the transaction status and the server will record the fall back to synchronous 2pc.
```

Comment:
Better to tell the indicator of fallback: `use_async_commit` of any lock is false.

Reply:
If it's 1pc, fallback is not detectable in lock resolving, because no locks will be written if 1pc fails.